lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/driver/DefaultVortexMaster.java

```diff
@@ -107,37 +111,78 @@ public void workerPreempted(final String id) {
     }
   }
 
-  /**
-   * Notify task completion to runningWorkers.
-   */
   @Override
-  public void taskletCompleted(final String workerId,
-                               final int taskletId,
-                               final Serializable result) {
-    runningWorkers.completeTasklet(workerId, taskletId, result);
+  public void workerReported(final String workerId, final WorkerReport workerReport) {
+    // TODO[JIRA REEF-942]: Fix when aggregation is allowed.
+
+    for (final TaskletReport taskletReport : workerReport.getTaskletReports()) {
+      switch (taskletReport.getType()) {
+      case TaskletResult:
+        final TaskletResultReport taskletResultReport = (TaskletResultReport) taskletReport;
+
+        final List<Integer> resultTaskletIds = taskletResultReport.getTaskletIds();
+        runningWorkers.doneTasklets(workerId, resultTaskletIds);
+        fetchDelegate(resultTaskletIds).completed(resultTaskletIds, taskletResultReport.getResult());
+
+        break;
+      case TaskletCancelled:
+        final TaskletCancelledReport taskletCancelledReport = (TaskletCancelledReport) taskletReport;
+        final List<Integer> cancelledIdToList = Collections.singletonList(taskletCancelledReport.getTaskletId());
+        runningWorkers.doneTasklets(workerId, cancelledIdToList);
+        fetchDelegate(cancelledIdToList).cancelled(taskletCancelledReport.getTaskletId());
+
+        break;
+      case TaskletFailure:
+        final TaskletFailureReport taskletFailureReport = (TaskletFailureReport) taskletReport;
+
+        final List<Integer> failureTaskletIds = taskletFailureReport.getTaskletIds();
+        runningWorkers.doneTasklets(workerId, taskletFailureReport.getTaskletIds());
+        fetchDelegate(failureTaskletIds).threwException(failureTaskletIds, taskletFailureReport.getException());
```

Comment:
We need to clarify `failureTaskletIds`. This's related to the comment I mentioned above.